controller.py

```diff
@@ -81,46 +83,157 @@
+        request.
 
-    if collection_details:
-        # A DataSource may be sent for collections with the
-        # ExternalIntegration.OPDS_IMPORT protocol.
-        data_source_name = request.args.get('data_source')
-        if data_source_name:
-            data_source_name = urllib.unquote(data_source_name)
+        This function is used by controller code and by a decorator
+        applied to Flask routes.
 
-        collection, ignore = Collection.from_metadata_identifier(
-            _db, collection_details, data_source=data_source_name
-        )
-        return collection
-    return None
+        :param required: True if a lack of authentication or bad
+                         authentication is an error condition.
 
+        :return: An IntegrationClient if one could be authenticated;
+                 otherwise None (if authentication is not required)
+                 or a ProblemDetail.
+        """
+        # Start with the assumption that there is no authenticated
+        # IntegrationClient.
+        flask.request.authenticated_client = None
+
+        header = flask.request.headers.get('Authorization')
+        if header:
+            if not 'bearer' in header.lower():
+                # No authentication mechanism other than a bearer token
+                # is supported
+                return INVALID_CREDENTIALS
+
+            # We can hopefully use the provided bearer token to
+            # authenticate an IntegrationClient.
+            try:
+                shared_secret = base64.b64decode(header.split(' ')[1])
+            except TypeError, e:
+                # The bearer token is ill-formed.
+                return INVALID_CREDENTIALS
+            client = IntegrationClient.authenticate(_db, shared_secret)
+
+            if client:
+                # Success!
+                flask.request.authenticated_client = client
+                return client
+            else:
+                # The bearer token doesn't identify any particular
+                # IntegrationClient.
+                return INVALID_CREDENTIALS
 
-class IndexController(object):
+        if required:
+            # No authentication was provided -- unfortuntely,
+            # it's required.
+            return INVALID_CREDENTIALS
+
+        # No credentials were provided, but authentication is optional.
+        return None
+
+
+class Controller(object):
+    """A generic controller superclass for the metadata wrangler."""
 
     def __init__(self, _db):
+        """Generic constructor.
+
+        :param _db: A scoped-session database connection.
+        """
         self._db = _db
+        self._default_collection_id = None
+
+    @property
+    def default_collection(self):
+        """Look up the 'unaffiliated' collection, which is used to register
+        identifiers referenced by unauthenticated clients in one-off tests.
+        """
+        if self._default_collection_id is None:
```

Comment:
Why was this changed from `getattr(self, ...` to `self._default ...`?

Reply:
_default_collection_id used to be set the first time it was used. Before that happened, `self._default_collection_id` would have raised an exception. Now it's set to `None` in the constructor, so we can always check the value, even if it turns out to be `None`.